src/psyclone/gocean1p0.py

```diff
@@ -1308,11 +1308,37 @@ def gen_ocl(self, parent):
         qlist = symtab.lookup_with_tag("opencl_cmd_queues").name
         flag = symtab.lookup_with_tag("opencl_error").name
 
-        # Then we call clEnqueueNDRangeKernel
-        parent.add(CommentGen(parent, " Launch the kernel"))
-        cnull = "C_NULL_PTR"
+        # Choose the command queue number to where to dispatch this kernel. We
+        # have do deal with possible dependencies to kernels dispatched in
+        # different command queues as the order in not guaranteed.
         queue_number = self._opencl_options['queue_number']
         cmd_queue = qlist + "({0})".format(queue_number)
+        outer_loop = self.parent.parent.parent.parent
+        dependency = outer_loop.backward_dependence()
+
+        # If the dependency is a Kernel add a barrier if the previous kernels
+        # was dispatched in a different command queue
+        if dependency and dependency.coded_kernels():
+            for kernel_dep in dependency.coded_kernels():
+                previous_queue = kernel_dep._opencl_options['queue_number']
+                if previous_queue != queue_number:
+                    # If the backward dependency is being executed in another
+                    # queue we add a barrier to make sure the previous kernel
+                    # has finished before this one starts.
+                    parent.add(AssignGen(
+                        parent,
+                        lhs=flag,
+                        rhs="clFinish({0}({1}))".format(
+                            qlist, str(previous_queue))))
+
+        # If the dependency is something other than a kernel, currently we
+        # dispatch everything else to queue 1, so add a barrier if this kernel
+        # is not on queue number 1.
```

Comment:
Done. I made a global for now but also added a TODO because ultimately this should be moved into the OCL transformation when this logic happens inside the transformation (and maybe even be configurable with a transformation option parameter)